File: app/src/main/java/com/xuie/android/ui/main/TransitionsFragment.java
package com.xuie.android.ui.main;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.app.ActivityOptions;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.AccelerateInterpolator;
import android.widget.Button;
import android.widget.ImageView;

import com.orhanobut.logger.Logger;
import com.xuie.android.R;

import java.util.Objects;

/**
 * A simple {@link Fragment} subclass.
 *
 * @author xuie
 */
public class TransitionsFragment extends Fragment {
    private Button makeSceneTransitionAnimation;
    private Button fabButton;
    private ImageView overShoot;
    private ImageView cirRevealDst;
    private ImageView cirRevealHypot;
    private ImageView cirRevealNormal;
    private Button explode;
    private Button slide;
    private Button fade;
    private Button cirRevealBtn;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_transitions, container, false);
        makeSceneTransitionAnimation = view.findViewById(R.id.make_scene_transition_animation);
        fabButton = view.findViewById(R.id.fab_button);
        overShoot = view.findViewById(R.id.over_shoot);
        cirRevealDst = view.findViewById(R.id.cir_reveal_dst);
        cirRevealHypot = view.findViewById(R.id.cir_reveal_hypot);
        cirRevealNormal = view.findViewById(R.id.cir_reveal_normal);
        explode = view.findViewById(R.id.explode);
        slide = view.findViewById(R.id.slide);
        fade = view.findViewById(R.id.fade);
        cirRevealBtn = view.findViewById(R.id.cir_reveal_btn);
        return view;
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        makeSceneTransitionAnimation.setOnClickListener(v -> {
            Intent intent = new Intent(getActivity(), OneActivity.class);
            ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(getActivity(), makeSceneTransitionAnimation, "share01");
            ActivityCompat.startActivity(Objects.requireNonNull(getActivity()), intent, options.toBundle());
        });

        fabButton.setOnClickListener(v -> {
            Logger.d("fab_button");
            Intent intent = new Intent(getActivity(), OneActivity.class);
            ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(getActivity(),
                    // 创建多个共享元素
                    Pair.create(makeSceneTransitionAnimation, "share01"),
                    Pair.create(fabButton, "share02")
            );

            ActivityCompat.startActivity(Objects.requireNonNull(getActivity()), intent, options.toBundle());
        });

        explode.setOnClickListener(v -> makeSceneTransitionAnimationNoParameter(0));
        slide.setOnClickListener(v -> makeSceneTransitionAnimationNoParameter(1));
        fade.setOnClickListener(v -> makeSceneTransitionAnimationNoParameter(2));

        overShoot.setOnClickListener(v -> {
            Logger.d("over_shoot");
            Intent intent = new Intent(getActivity(), TransitionsObjectActivity.class);
            ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(getActivity(), overShoot, "shareOverShoot");
            ActivityCompat.startActivity(Objects.requireNonNull(getActivity()), intent, options.toBundle());
        });


        cirRevealBtn.setOnClickListener(v -> {
            // http://blog.jobbole.com/77015/ 圆形显示
            if (cirRevealDst.getVisibility() != View.VISIBLE) {
                Animator anim = ViewAnimationUtils.createCircularReveal(
                        cirRevealDst,
                        cirRevealDst.getWidth() / 2,
                        cirRevealDst.getHeight() / 2,
                        0,
                        Math.max(cirRevealDst.getWidth(), cirRevealDst.getHeight()));

                anim.addListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationStart(Animator animation) {
                        super.onAnimationStart(animation);
                        cirRevealDst.setVisibility(View.VISIBLE);
                    }
                });

                anim.start();
            } else {
                Animator anim = ViewAnimationUtils.createCircularReveal(
                        cirRevealDst,
                        cirRevealDst.getWidth() / 2,
                        cirRevealDst.getHeight() / 2,
                        Math.max(cirRevealDst.getWidth(), cirRevealDst.getHeight()),
                        0);

                anim.addListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        super.onAnimationEnd(animation);
                        cirRevealDst.setVisibility(View.INVISIBLE);
                    }
                });

                anim.start();
            }
        });

        cirRevealNormal.setOnClickListener(v -> {
            Animator mNormalAnimator = ViewAnimationUtils.createCircularReveal(
                    cirRevealNormal,
                    cirRevealNormal.getWidth() / 2,
                    cirRevealNormal.getHeight() / 2,
                    0,
                    Math.max(cirRevealNormal.getWidth(), cirRevealNormal.getHeight()));
            mNormalAnimator.setInterpolator(new AccelerateInterpolator());
            mNormalAnimator.start();
        });

        cirRevealHypot.setOnClickListener(v -> {
            //勾股定理得到斜边长度
            float endRadius = (float) Math.hypot(cirRevealHypot.getWidth(),
                    cirRevealHypot.getHeight());
            Animator mHypotAnimator = ViewAnimationUtils.createCircularReveal(cirRevealHypot,
                    cirRevealHypot.getWidth(), 0, 0, endRadius);
            mHypotAnimator.setInterpolator(new AccelerateDecelerateInterpolator());
            mHypotAnimator.start();
        });
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    private void makeSceneTransitionAnimationNoParameter(int flag) {
        Intent intent = new Intent(getActivity(), OneActivity.class);
        intent.putExtra("flag", flag);
        ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(getActivity());
        ActivityCompat.startActivity(Objects.requireNonNull(getActivity()), intent, options.toBundle());
    }
}

File: app/src/main/java/com/xuie/android/ui/main/TransitionsFragment.kt
package com.xuie.android.ui.main

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.app.ActivityOptions
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.AccelerateInterpolator
import android.widget.Button
import android.widget.ImageView
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.xuie.android.R
import com.xuie.android.util.log
import java.util.*

/**
 * A simple [Fragment] subclass.
 *
 * @author xuie
 */
class TransitionsFragment : Fragment() {
  private var makeSceneTransitionAnimation: Button? = null
  private var fabButton: Button? = null
  private var overShoot: ImageView? = null
  private var cirRevealDst: ImageView? = null
  private var cirRevealHypot: ImageView? = null
  private var cirRevealNormal: ImageView? = null
  private var explode: Button? = null
  private var slide: Button? = null
  private var fade: Button? = null
  private var cirRevealBtn: Button? = null

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
    val view = inflater.inflate(R.layout.fragment_transitions, container, false)
    makeSceneTransitionAnimation = view.findViewById(R.id.make_scene_transition_animation)
    fabButton = view.findViewById(R.id.fab_button)
    overShoot = view.findViewById(R.id.over_shoot)
    cirRevealDst = view.findViewById(R.id.cir_reveal_dst)
    cirRevealHypot = view.findViewById(R.id.cir_reveal_hypot)
    cirRevealNormal = view.findViewById(R.id.cir_reveal_normal)
    explode = view.findViewById(R.id.explode)
    slide = view.findViewById(R.id.slide)
    fade = view.findViewById(R.id.fade)
    cirRevealBtn = view.findViewById(R.id.cir_reveal_btn)
    return view
  }

  @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    makeSceneTransitionAnimation!!.setOnClickListener { v ->
      val intent = Intent(activity, OneActivity::class.java)
      val options = ActivityOptions.makeSceneTransitionAnimation(activity, makeSceneTransitionAnimation, "share01")
      ActivityCompat.startActivity(Objects.requireNonNull<FragmentActivity>(activity), intent, options.toBundle())
    }

    fabButton!!.setOnClickListener { v ->
      log { "fab_button" }
      val intent = Intent(activity, OneActivity::class.java)
      val options = ActivityOptions.makeSceneTransitionAnimation(activity,
          // 创建多个共享元素
          Pair.create<View, String>(makeSceneTransitionAnimation, "share01"),
          Pair.create<View, String>(fabButton, "share02")
      )

      ActivityCompat.startActivity(Objects.requireNonNull<FragmentActivity>(activity), intent, options.toBundle())
    }

    explode!!.setOnClickListener { v -> makeSceneTransitionAnimationNoParameter(0) }
    slide!!.setOnClickListener { v -> makeSceneTransitionAnimationNoParameter(1) }
    fade!!.setOnClickListener { v -> makeSceneTransitionAnimationNoParameter(2) }

    overShoot!!.setOnClickListener { v ->
      log { "over_shoot" }
      val intent = Intent(activity, TransitionsObjectActivity::class.java)
      val options = ActivityOptions.makeSceneTransitionAnimation(activity, overShoot, "shareOverShoot")
      ActivityCompat.startActivity(Objects.requireNonNull<FragmentActivity>(activity), intent, options.toBundle())
    }


    cirRevealBtn!!.setOnClickListener { v ->
      // http://blog.jobbole.com/77015/ 圆形显示
      if (cirRevealDst!!.visibility != View.VISIBLE) {
        val anim = ViewAnimationUtils.createCircularReveal(
            cirRevealDst,
            cirRevealDst!!.width / 2,
            cirRevealDst!!.height / 2,
            0f,
            Math.max(cirRevealDst!!.width, cirRevealDst!!.height).toFloat())

        anim.addListener(object : AnimatorListenerAdapter() {
          override fun onAnimationStart(animation: Animator) {
            super.onAnimationStart(animation)
            cirRevealDst!!.visibility = View.VISIBLE
          }
        })

        anim.start()
      } else {
        val anim = ViewAnimationUtils.createCircularReveal(
            cirRevealDst,
            cirRevealDst!!.width / 2,
            cirRevealDst!!.height / 2,
            Math.max(cirRevealDst!!.width, cirRevealDst!!.height).toFloat(),
            0f)

        anim.addListener(object : AnimatorListenerAdapter() {
          override fun onAnimationEnd(animation: Animator) {
            super.onAnimationEnd(animation)
            cirRevealDst!!.visibility = View.INVISIBLE
          }
        })

        anim.start()
      }
    }

    cirRevealNormal!!.setOnClickListener { v ->
      val mNormalAnimator = ViewAnimationUtils.createCircularReveal(
          cirRevealNormal,
          cirRevealNormal!!.width / 2,
          cirRevealNormal!!.height / 2,
          0f,
          Math.max(cirRevealNormal!!.width, cirRevealNormal!!.height).toFloat())
      mNormalAnimator.interpolator = AccelerateInterpolator()
      mNormalAnimator.start()
    }

    cirRevealHypot!!.setOnClickListener { v ->
      //勾股定理得到斜边长度
      val endRadius = Math.hypot(cirRevealHypot!!.width.toDouble(),
          cirRevealHypot!!.height.toDouble()).toFloat()
      val mHypotAnimator = ViewAnimationUtils.createCircularReveal(cirRevealHypot,
          cirRevealHypot!!.width, 0, 0f, endRadius)
      mHypotAnimator.interpolator = AccelerateDecelerateInterpolator()
      mHypotAnimator.start()
    }
  }

  @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
  private fun makeSceneTransitionAnimationNoParameter(flag: Int) {
    val intent = Intent(activity, OneActivity::class.java)
    intent.putExtra("flag", flag)
    val options = ActivityOptions.makeSceneTransitionAnimation(activity)
    ActivityCompat.startActivity(Objects.requireNonNull<FragmentActivity>(activity), intent, options.toBundle())
  }
}
